marshal/admin.go

```diff
@@ -0,0 +1,322 @@
+		return false
+	}
+
+	// Next, try to claim the partition.
+	if !a.marshaler.ClaimPartition(topic, partID) {
+		log.Errorf("[%s:%d] Admin couldn't claim partition to set Kafka offset",
+			topic, partID)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	// Continuously heartbeat the last offsets.
+	a.addClaimAttempt(topic, partID, offsets.Current, newOffset)
+	go a.heartbeatLoop(topic, partID, offsets.Current, stopHeartbeat)
+	return true
+}
+
+// constructReleaseGroupMessage constructs a ReleaseGroup message to write to the marshal topic.
+func (a *consumerGroupAdmin) constructReleaseGroupMessage() *msgReleaseGroup {
+	now := time.Now()
+	base := &msgBase{
+		Time:       int(now.Unix()),
+		InstanceID: newInstanceID(),
+		ClientID:   a.clientID,
+		GroupID:    a.groupID,
+	}
+	return &msgReleaseGroup{
+		msgBase:       *base,
+		MsgExpireTime: int(now.Add(a.pauseTimeout).Unix()),
+	}
+}
+
+// pauseGroupAndWaitForRelease is called for every partition we'd like to change the offset for.
+// It first sends a ReleaseGroup message to that partition, then waits for it to be released,
+// then attempts to claim it.
+func (a *consumerGroupAdmin) pauseGroupAndWaitForRelease(topicName string, partID int) bool {
+	topic := a.marshaler.cluster.getPartitionState(a.groupID, topicName, partID)
+	rg := a.constructReleaseGroupMessage()
+	_, err := a.marshaler.cluster.producer.Produce(MarshalTopic,
+		int32(topic.claimPartition), &proto.Message{Value: []byte(rg.Encode())})
+	if err != nil {
+		log.Errorf("[%s:%d] Admin failed to produce ReleaseMessage group to Kafka: %s",
+			topicName, partID, err)
+		return false
+	}
+
+	// Wait for the paused consumer to release its claim.
+	partitionClaim := a.marshaler.GetPartitionClaim(topicName, partID)
+
+	start := time.Now()
+	for partitionClaim.isClaimed(time.Now().Unix()) {
+		if time.Since(start) > consumerReleaseClaimWaitTime {
+			log.Errorf("[%s:%d] Admin timed out waiting for consumer to release its claim",
+				topic, partID)
+			return false
+		}
+
+		time.Sleep(consumerReleaseClaimWaitSleep)
+		partitionClaim = a.marshaler.GetLastPartitionClaim(topicName, partID)
+	}
+
+	return true
+}
+
+// SetConsumerGroupPosition sets where the consumer group identified by groupID
+// should start reading from for given partitions.
+func (a *consumerGroupAdmin) SetConsumerGroupPosition(groupID string,
+	offsets map[string]map[int]int64) error {
+
+	log.Infof("Admin %s going to pause consumer group %s", a.clientID, groupID)
+	// Send out a ReleaseGroup message to Marshal for each partition we want to set the position for,
+	// then wait for all the partitions to be released.
+	pauseResults := make(chan bool)
+	defer close(pauseResults)
+	numClaims := 0
+	for topicName, partitionOffsets := range offsets {
+		for partID, _ := range partitionOffsets {
+			numClaims++
+			go func(topicName string, partID int) {
+				pauseResults <- a.pauseGroupAndWaitForRelease(topicName, partID)
```

Comment:
This is still pausing once per (topic, partition) in the checkpoint. It needs to pause only once per **Marshal** (topic, partition) which is a different operation.